Inspect/Test File Checking Performance

Booksbaum opened the pull request:
Even with newest changes performance while writing code in bigger projects (like in FSAC) is still unpleasant: Still lots of small stutter.


So: time to test some things:

Issue is while writing -> issue is most likely in [`checkFile`](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L251-L262)

<br/>

First test: disable checking of related files in [`Commands.CheckFileAndAllDependentFilesInAllProjects`](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete.Core/Commands.fs#L764). Less checking should result in better performance.  
But: doesn't have huge impact. I think slightly better, but just barely.

<br/>

Next: change [debounce timeout](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L295) for `checkChangeFiles` (from `250ms` to `350ms`).  
Now writing does work without any mayor lags (but with errors displayed slightly later)


`checkChangedFile` in [debouncer](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete.Core/Utils.fs#L731) gets [started and immediately forgotten](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete.Core/Utils.fs#L742) (-> `checkChangedFile` get [started](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L293) with `Async.Start`).  
So it seems: multiple `checkFile` can run at the same time.  
Logging proofs that (with default timeout) (number after path is file version):
```
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 38, dontCheckRelatedFiles=False, since last check=00:00:00.7844834)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 42, dontCheckRelatedFiles=False, since last check=00:00:00.5009282)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 44, dontCheckRelatedFiles=False, since last check=00:00:00.5212574)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 24) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 32) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 49, dontCheckRelatedFiles=False, since last check=00:00:00.7109144)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 50, dontCheckRelatedFiles=False, since last check=00:00:00.4989457)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 44) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 50) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 38) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 49) finished
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 42) finished
```
-> multiple `checkFile`s running at same time. (also: finish out of order)

<br/>

So next test: wait in debouncer till `checkFile` finished (-> `do! fn arg.Value` instead of [just calling](https://github.com/fsharp/FsAutoComplete/blob/680a95b9d02c330d5b8a3a459271cc532efb368c/src/FsAutoComplete.Core/Utils.fs#L742) `fn`).  
This too made writing without mayor stutter possible. But again with errors display a bit later.  
It behaves quite similar to changing debounce timeout. Which is expected: larger debounce timeout -> less likely for multiple `checkFile`s to run at the same time


</br>
</br>

How much later (and less often) does `checkFile` get called when we ensure just one `checkFile` is running at the same time?


Fire & forget (`Async |> Start`):
```
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 58, dontCheckRelatedFiles=False, since last check=00:00:00.4630319)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 63, dontCheckRelatedFiles=False, since last check=00:00:00.7380374)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 64, dontCheckRelatedFiles=False, since last check=00:00:00.6643221)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 66, dontCheckRelatedFiles=False, since last check=00:00:00.6950823)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 70, dontCheckRelatedFiles=False, since last check=00:00:00.5220683)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 71, dontCheckRelatedFiles=False, since last check=00:00:00.6024039)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 73, dontCheckRelatedFiles=False, since last check=00:00:00.3460527)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 78, dontCheckRelatedFiles=False, since last check=00:00:00.8328280)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 84, dontCheckRelatedFiles=False, since last check=00:00:00.6693897)
```

wait (`do!`)
```
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 101, dontCheckRelatedFiles=False, since last check=00:00:03.0156481)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 114, dontCheckRelatedFiles=False, since last check=00:00:02.4365100)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 125, dontCheckRelatedFiles=False, since last check=00:00:02.0815498)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 129, dontCheckRelatedFiles=False, since last check=00:00:01.8309675)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 137, dontCheckRelatedFiles=False, since last check=00:00:01.4969826)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 147, dontCheckRelatedFiles=False, since last check=00:00:02.1179966)
[LSP] checkFile([...]\src\FsAutoComplete\FsAutoComplete.Lsp.fs, 158, dontCheckRelatedFiles=False, since last check=00:00:02.2869400)
```

-> now checks are seconds apart instead of under a second.
(Similar behavior for increased debounce timeout)


Though later & less diagnostic updates are unfortunate, for me that's worth the change as it removes the greater annoyance of stutter.


------------------

This PR introduces some debug settings to change the different behaviours mentioned above. Without any of the settings specified, no behaviour is changed and FSAC behaves just like current `main`.  
(-> for testing, not to be merged)

* [`FSharp.debug.dontCheckRelatedFiles`](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L277-L289): when enabled just checks current file and no related source files
* [`FSharp.debug.checkFileDebouncerTimeout`](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete.Core/Utils.fs#L758): debouncer timeout in ms: how long should there no be source change before `checkFile` gets called
  * Note: no validation of correct number. Probably throws exception with negative number.
* [`FSharp.debug.waitTillFileChecked`](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete.Core/Utils.fs#L743-L746): only one `checkFile` can run at same time
* [`FSharp.debug.logDurationBetweenCheckFiles`](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L261-L275): Logs durations between two consecutive `checkFile` calls (-> output above)
  * With log level `Warn` -> visible in VSCode console even without any additional log setting.
* [`FSharp.debug.logCheckFileFinished`](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L293-L298): Logs `checkFile` finished.  
  In combination with `logDurationBetweenCheckFiles`: can trace start and end of `checkFile` (see output above)


To set settings in VsCode & Ionide: specify in `settings.json`. Settings are then immediately adopted and reflected by FSAC (-> can be changed while running. No restart necessary)


<br/>
<br/>

--------------------

<br/>

I don't know if any default behaviour should be changed based on these findings. But i would very much like to have the option to prevent simultaneous file checking (`waitTillFileChecked`) as it greatly increased performance for me.

Additional: as logs above show, multiple simultaneous running checkFiles indicate a bug in FSAC: different checks on different file states finishing at different times in different orders.
So limit to just one checkFile at a time might actually increase correctness
-> maybe make that the default?

(maybe even with further adjustments like cancelling a running file check when a new change comes in? (-> less diag updates for client, but faster diags when user done writing))

Booksbaum committed:
Format Code

Booksbaum committed:
Add debug settings for `checkFile`

`FSharp.debug.dontCheckRelatedFiles`,
`FSharp.debug.checkFileDebouncerTimeout`,
`FSharp.debug.waitTillFileChecked`,
`FSharp.debug.logDurationBetweenCheckFiles`,
`FSharp.debug.logCheckFileFinished`

baronfel commented:
I haven't looked in detail (yet!) but I can say that the intent is to only have on concurrent CheckFile _per file_ at a time. So we can assume that should be the default.

baronfel commented on src/FsAutoComplete.Core/Utils.fs:
```diff
@@ -728,28 +728,49 @@ module Version =
 //source: https://nbevans.wordpress.com/2014/08/09/a-simple-stereotypical-javascript-like-debounce-service-for-f/
-type Debounce<'a>(timeout, fn) =
-  let debounce fn timeout =
+type Debounce<'a>(timeout, fn) as x =
```
curious about the `as x` pattern here - I see it used below in the `mailbox`, is that required because `mailbox` became a value?

Booksbaum replied:
`MyType(...) as x` -> binds `x` as `this` identifier. Basically just like `member x.MyMethod(...)=..`, but here `x` is just valid for current method, while `x` in `MyType(...) as x` is valid for the whole class -- including outside of members (like inside let bindings).  

`let mailbox = ...` isn't a member or inside a member, but [calls a class member](https://github.com/Booksbaum/FsAutoComplete/blob/07629d1025541490ed3dd7f037eb4ec5f02cdb6b/src/FsAutoComplete.Core/Utils.fs#L743) -> needs access to `this` (here named `x`). Wasn't necessary previously because no member was called (-> no access to `this` was necessary).

baronfel replied:
ok, so if we make `WaitForFnToFinish` the default (as I believe it should be), then the 'self' identifier could be removed entirely.

Booksbaum replied:
Yes

I just introduced the property `WaitForFnToFinish` for testing -> to change on the fly how it should behave (`FSharp.debug.waitTillFileChecked`)

baronfel commented:
> (maybe even with further adjustments like cancelling a running file check when a new change comes in? (-> less diag updates for client, but faster diags when user done writing))

100% agree with this as well, it would just be more bookkeeping I suppose.

Booksbaum committed:
Cancel `checkFile` when new `checkFile` should be executed

Result: for large files/project: in practice: cancels all but last `checkFile` execution

Note: Doesn't use `FSharp.debug.checkFileDebouncerTimeout` and `FSharp.debug.waitTillFileChecked`

Booksbaum commented:
>> (maybe even with further adjustments like cancelling a running file check when a new change comes in? (-> less diag updates for client, but faster diags when user done writing))
>
> 100% agree with this as well, it would just be more bookkeeping I suppose.

Newest commit adds that behaviour.  
Actually quite easy to implement [with Observable](https://github.com/Booksbaum/FsAutoComplete/blob/574f5819affabf294faba4b425063db324500f24/src/FsAutoComplete/FsAutoComplete.Lsp.fs#L346-L351). Seems to work just fine.

<br/>

But I'm not sure we should use this: in practice (with large files/projects -> long checking duration) it prevents most diags while writing: though `CheckFile` gets started quite often, it usually gets cancelled soon after (and a new `CheckFile` gets started).

-> vs. without cancelling: occasional diags while writing, but final diags slightly later (because prev `checkFile` must first complete)     
I personally prefer this behaviour.

baronfel commented:
You make a compelling point. Here's where I think we should go with this:

* remove the cancellation support
* default `waitTillFileChecked` to `true` or remove that knob entirely in favor of always awaiting - that was always the _intended_ behavior of the debouncing, and you fixed a major perf bug
* the other defaults I'm perfectly ok with. I can add them to the Ionide debug options configuration docs, or you can on a subsequent PR.

All in all this is a lovely set of changes :)

Booksbaum committed:
Change: always allow just one running `checkFile`

Remove `waitTillFileChecked` setting

Remove cancelling of `checkFile`

Booksbaum committed:
Add Debug Settings to Readme

Booksbaum committed:
Update `checkFile` logging

Rename `LogCheckFileFinished` to `LogCheckFileDuration`

Measure duration of `checkFile`

Initialize checkFileDebouncer with default CheckFileDebouncerTimeout

Booksbaum committed:
Remove unused opens

Booksbaum committed:
Format Code

Booksbaum commented:
Just one running `checkFile` without cancellation is now the only option, both other options (cancellation & parallel) were removed.

<br/>

Available settings are now: `FSharp.Debug.DontCheckRelatedFiles`, `FSharp.Debug.CheckFileDebouncerTimeout`, `FSharp.Debug.LogDurationBetweenCheckFiles`, `FSharp.Debug.LogCheckFileDuration`; all others were removed.  

I currently log debug setting changes & duration as Warning. Though their content match info or verbose level, I prefer to keep Warning level: visible in VSCode without any extra verbose logging setting and easier to isolate in log list. And duration logs must be explicitly enabled and are not part of usual logging messages -> I think Warning is ok

<br/>

> I can add them to the Ionide debug options configuration docs, or you can on a subsequent PR.

I added the settings to [FSAC Readme](https://github.com/Booksbaum/FsAutoComplete/blob/b4de270093bb46fab0f739805a8152901e2428a8/README.md#debug-settings). I don't think we have to additional add these to Ionide docs as they are only used for Debugging and Testing of FSAC.

baronfel commented:
Awesome, I think this makes a ton of sense. I'll poke around a bit, but I'm happy to merge this soon :)
